core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableInterpretable.java

```diff
@@ -137,9 +148,29 @@ static Bindable getBindable(ClassDeclaration expr, String s, int fieldCount)
       // Add line numbers to the generated janino class
       cbe.setDebuggingInformation(true, true, true);
     }
+
+    if (BINDABLE_CACHE_MAX_SIZE != 0) {
+      StaticFieldDetector detector = new StaticFieldDetector();
+      expr.accept(detector);
+      if (!detector.containsStaticField) {
+        return BINDABLE_CACHE.get(s, () -> (Bindable) cbe.createInstance(new StringReader(s)));
```

Comment:
Do you intend to cache classes or instances?
If you cache instances, then you rule out `non-static` fields as well otherwise two threads might happen to mutate the same object concurrently.

So we should either ensure the generated objects are reusable, or (which I like better) cache just the class or even factory for bindable classes.

In other words, if generated class implements `Supplier<Binable>` as a factory method, then we could cache "factory" in the BINDABLE_CACHE, and call `BINDABLE_CACHE.get(...).get()` to create unshared objects.

Reply:
I like the fact that you are talking about caching factories because this is more or less my intention from a slight different persepctive. 

I was thinking of factories of `Enumerable` objects that's why I am putting in the cache instances of the `Bindable` interface. 

For me the "factory" is the `Bindable` instance. Different threads can call the `bind` method with a different `DataContext` to obtain a new `Enumerable` object so that's why I want `Bindable` objects to be thread-safe (see [CALCITE-2701](https://issues.apache.org/jira/browse/CALCITE-2701)).

Reply:
@zabetak ,I think you should cache the code of this query,rather than the instance.
If some one introduce fields into this class,the instance will not be  thread-safe

Reply:
@zabetak  After see [CALCITE-2701](https://issues.apache.org/jira/browse/CALCITE-2701), I think it is the right way. But if we can cache the whole plan for a query, is it beneficial for low latency?